oak/server/storage/storage_node.cc

```diff
@@ -18,42 +17,102 @@
 #include "oak/server/storage/storage_node.h"
 
 #include "absl/memory/memory.h"
+#include "absl/strings/escaping.h"
+#include "asylo/util/cleansing_types.h"
 #include "asylo/util/logging.h"
 #include "grpcpp/create_channel.h"
 #include "oak/common/app_config.h"
 #include "oak/proto/storage_channel.pb.h"
 
 namespace oak {
 
+constexpr size_t kMaxMessageSize = 1 << 16;
+
 static std::string GetStorageId(const std::string& storage_name) {
   // TODO: Generate name-based UUID.
   return storage_name;
 }
 
+static asylo::CleansingVector<uint8_t> GetStorageEncryptionKey(const std::string& storage_name) {
+  // TODO: Request encryption key from escrow service.
+  std::string encryption_key =
+      absl::HexStringToBytes("c0dedeadc0dedeadc0dedeadc0dedeadc0dedeadc0dedeadc0dedeadc0dedead");
+  return asylo::CleansingVector<uint8_t>(encryption_key.begin(), encryption_key.end());
+}
+
 StorageNode::StorageNode(const std::string& name, const std::string& storage_address)
-    : NodeThread(name) {
+    : NodeThread(name),
+      fixed_nonce_generator_(new oak::FixedNonceGenerator()),
+      datum_name_cryptor_(kMaxMessageSize, fixed_nonce_generator_),
+      datum_value_cryptor_(kMaxMessageSize, new asylo::AesGcmSivNonceGenerator()) {
   storage_service_ = oak::Storage::NewStub(
       grpc::CreateChannel(storage_address, grpc::InsecureChannelCredentials()));
 }
 
+const asylo::StatusOr<std::string> StorageNode::EncryptDatumName(
+    const absl::string_view& datum_name) {
+  fixed_nonce_generator_->set_datum_name(datum_name);
+  return Encrypt(&datum_name_cryptor_, datum_name);
+}
+
+const asylo::StatusOr<std::string> StorageNode::EncryptDatumValue(
+    const absl::string_view& datum_value) {
+  return Encrypt(&datum_value_cryptor_, datum_value);
+}
+
+const asylo::StatusOr<std::string> StorageNode::Encrypt(asylo::AesGcmSivCryptor* cryptor,
+                                                        const absl::string_view& data) {
+  asylo::CleansingVector<uint8_t> key = GetStorageEncryptionKey("foo");
```

Comment:
Some sort of TODO: replace foo with... here?

Reply:
Done.